src/coreclr/vm/method.cpp

```diff
@@ -4099,3 +4099,694 @@ void MethodDesc::PrepareForUseAsAFunctionPointer()
+                case CEE_BLT:
+                case CEE_BNE_UN:
+                case CEE_BGE_UN:
+                case CEE_BGT_UN:
+                case CEE_BLE_UN:
+                case CEE_BLT_UN:
+                    SkipIL(4);
+                    break;
+                case CEE_SWITCH:
+                    {
+                        uint32_t count = ReadILUInt32();
+                        if (count > 0x1FFFFFFF)
+                            ThrowHR(COR_E_BADIMAGEFORMAT);
+                        SkipIL(count * 4);
+                    }
+                    break;
+                default:
+                    continue;
+            }
+        }
+
+        ClrSafeInt<COUNT_T> dataStreamSize;
+        DWORD nonCodeAlternateDataSize = 0;
+        DWORD alternateNonTypeRefStreamSize = 0;
+        PVOID nonCodeAlternateData = nonCodeAlternateBlob.GetSignature(&nonCodeAlternateDataSize);
+        PVOID alternateNonTypeRefStreamData = alternateNonTypeRefStream.GetSignature(&alternateNonTypeRefStreamSize);
+        dataStreamSize = ClrSafeInt<COUNT_T>(S_SIZE_T(nonCodeAlternateDataSize) + S_SIZE_T(ilStream.GetCount()) + S_SIZE_T(alternateNonTypeRefStreamSize));
+        if (dataStreamSize.IsOverflow())
+            ThrowHR(COR_E_BADIMAGEFORMAT);
+
+        uint8_t *pData = pDataStream->OpenRawBuffer(dataStreamSize.Value());
+        memcpy(pData, nonCodeAlternateData, nonCodeAlternateDataSize);
+        ilStream.Copy(pData + nonCodeAlternateDataSize, ilStream.Begin(), ilStream.GetCount());
+        memcpy(pData + nonCodeAlternateDataSize + ilStream.GetCount(), alternateNonTypeRefStreamData, alternateNonTypeRefStreamSize);
+        pDataStream->CloseRawBuffer();
+    }
+
+    // Convert a token which may be a memberref parent into a member ref coded index
+    uint32_t DecodeMemberRefCodedIndexToken(uint32_t codedIndex)
+    {
+        static const mdToken s_tableTokenTypes[] = {mdtTypeDef, mdtTypeRef, mdtModuleRef, mdtMethodDef, mdtTypeSpec, 0, 0, 0};
+        return TokenFromRid((codedIndex & 0xFFFFFF8) >> 3, s_tableTokenTypes[codedIndex & 0x7]);
+    }
+
+    uint32_t MemberRefParentCodedIndex(mdToken tk)
+    {
+        RID     rid = RidFromToken(tk);
+        ULONG32 ulTyp = TypeFromToken(tk);
+
+        if ((rid > 0xFFFFFF) || rid == 0)
+        {
+            ThrowHR(COR_E_BADIMAGEFORMAT);
+        }
+
+        rid = (rid << 3);
+
+        // TypeDef  is encoded with low bits 000
+        // TypeRef  is encoded with low bits 001
+        // ModuleRef is encoded with low bits 010
+        // MethodDef is encoded with low bit 011
+        // TypeSpec is encoded with low bits 100
+        for (uint32_t codedIndexType = 0; codedIndexType <= 7; codedIndexType++)
+        {
+            if (DecodeMemberRefCodedIndexToken(rid + codedIndexType) == tk)
+            {
+                return rid + codedIndexType;
+            }
+        }
+        ThrowHR(COR_E_BADIMAGEFORMAT);
+    }
+
+    uint32_t GetAlternateToken(uint32_t inputToken)
```

Comment:
It might be useful to mention a reference to the equivalent C# code in a code comment somewhere around here.